File: crates/klio-stdlib/src/pack_builder.rs
//! Build a `.klio-pack` byte stream describing this crate's stdlib
//! surface. Used by `klio pack stdlib` and by the `klio-stdlib-pack`
//! build script to embed a pack inside the interpreter binary.
//!
//! The MVP form emits `manifest`, `symbols`, and `bindings` sections
//! only; the AST / resolved / typeck sections are reserved for the
//! interpreted-stdlib path that follows after the MVP.

use std::path::{Path, PathBuf};

use klio_pack::schema::{
    encode, Binding, BindingKind, BindingManifest, ModifierBits, PackManifest, Purity, SourceBundle,
    SourceFile, SourceLoc, SymbolIndex, SymbolKind, SymbolRecord,
};
use klio_pack::{section_names, Compression, PackError, PackWriter};

use crate::{generated, implementation, param_names, IMPLICITLY_IMPORTED_PACKAGES};

/// Curated set of upstream stdlib commonMain `.kt` files the embedded
/// stdlib pack ships verbatim as a `SOURCES` section, so the
/// interpreter consumes the real upstream Kotlin instead of (or
/// alongside) the mined Rust surface.
///
/// This is the general, include-list-driven mechanism: each entry is a
/// path **relative to `kotlin/libraries/stdlib`** in the local upstream
/// Kotlin checkout. The list is intentionally scoped to `kotlin.time`
/// for now; growing the consumed-from-source surface later is a matter
/// of appending entries here (plus authoring any matching `actual`s).
///
/// The platform `actual`s for the `internal expect` declarations in
/// these files (and the one `public expect enum class DurationUnit`)
/// are supplied by [`KLIO_STDLIB_ACTUAL_FILES`], authored under
/// `crates/klio-stdlib/`.
const CURATED_UPSTREAM_SOURCES: &[&str] = &[
    "src/kotlin/time/Duration.kt",
    "src/kotlin/time/DurationUnit.kt",
    "src/kotlin/time/longSaturatedMath.kt",
    "src/kotlin/time/measureTime.kt",
    "src/kotlin/time/TimeSource.kt",
    "src/kotlin/time/TimeSources.kt",
    "src/kotlin/time/Clock.kt",
    "src/kotlin/time/Clocks.kt",
    "src/kotlin/time/Instant.kt",
    "src/kotlin/time/ExperimentalTime.kt",
    "src/kotlin/coroutines/CoroutineContext.kt",
    "src/kotlin/coroutines/CoroutineContextImpl.kt",
    "src/kotlin/coroutines/ContinuationInterceptor.kt",
    "src/kotlin/coroutines/Continuation.kt",
    "common/src/generated/_Strings.kt",
    "common/src/generated/_Maps.kt",
    "common/src/generated/_Collections.kt",
    "common/src/generated/_Sets.kt",
    "common/src/generated/_Ranges.kt",
    "common/src/generated/_Sequences.kt",
    "common/src/kotlin/Comparator.kt",
    "src/kotlin/util/Standard.kt",
    "src/kotlin/util/Tuples.kt",
    "src/kotlin/util/Result.kt",
    "src/kotlin/util/Numbers.kt",
    "src/kotlin/util/Lateinit.kt",
    "src/kotlin/util/FloorDivMod.kt",
    "src/kotlin/util/Suspend.kt",
    "src/kotlin/util/DeepRecursive.kt",
    "src/kotlin/util/HashCode.kt",
    "src/kotlin/util/Preconditions.kt",
    "src/kotlin/collections/AbstractCollection.kt",
    "src/kotlin/collections/AbstractList.kt",
    "src/kotlin/collections/AbstractMap.kt",
    "src/kotlin/collections/AbstractSet.kt",
    "src/kotlin/collections/Collections.kt",
    "src/kotlin/collections/Iterables.kt",
    "src/kotlin/collections/Maps.kt",
    "src/kotlin/collections/Iterators.kt",
    "src/kotlin/collections/MapAccessors.kt",
    "src/kotlin/collections/SlidingWindow.kt",
    "src/kotlin/collections/MutableCollections.kt",
    "src/kotlin/collections/ReversedViews.kt",
    "src/kotlin/collections/Grouping.kt",
    "src/kotlin/collections/IndexedValue.kt",
    "src/kotlin/collections/MapWithDefault.kt",
    "src/kotlin/collections/Sequence.kt",
    "src/kotlin/collections/SequenceBuilder.kt",
    "src/kotlin/collections/Sets.kt",
    "src/kotlin/comparisons/Comparisons.kt",
    "src/kotlin/comparisons/compareTo.kt",
    "src/kotlin/annotations/OptIn.kt",
    "src/kotlin/annotations/Multiplatform.kt",
    "src/kotlin/annotations/ExperimentalStdlibApi.kt",
    "src/kotlin/annotations/Inference.kt",
    "src/kotlin/annotations/Throws.kt",
    "src/kotlin/annotations/WasExperimental.kt",
    "src/kotlin/annotations/ReturnValue.kt",
    "src/kotlin/annotations/ConsistentCopyVisibility.kt",
    "src/kotlin/annotations/VersionOverloads.kt",
    "src/kotlin/properties/Interfaces.kt",
    "src/kotlin/properties/ObservableProperty.kt",
    "src/kotlin/properties/PropertyReferenceDelegates.kt",
    "src/kotlin/text/CharacterCodingException.kt",
    "src/kotlin/io/encoding/ExperimentalEncodingApi.kt",
    "src/kotlin/uuid/ExperimentalUuidApi.kt",
    "src/kotlin/experimental/ExperimentalNativeApi.kt",
    "src/kotlin/concurrent/Volatile.kt",
    "src/kotlin/concurrent/atomics/ExperimentalAtomicApi.kt",
    "src/kotlin/contextParameters/ExperimentalContextParameters.kt",
    "src/kotlin/ranges/Range.kt",
    "src/kotlin/ranges/Ranges.kt",
    "src/kotlin/ranges/Progressions.kt",
    "src/kotlin/ranges/ProgressionIterators.kt",
    "src/kotlin/AutoCloseable.kt",
    "src/kotlin/contextParameters/Context.kt",
    "src/kotlin/contextParameters/ContextOf.kt",
    "src/kotlin/properties/Delegates.kt",
    "src/kotlin/Function.kt",
    "src/kotlin/random/URandom.kt",
    "src/kotlin/enums/EnumEntries.kt",
    "src/kotlin/Library.kt",
    "src/kotlin/uuid/Uuid.kt",
    "src/kotlin/Unit.kt",
    "src/kotlin/Throwable.kt",
    "src/kotlin/text/regex/MatchResult.kt",
    "src/kotlin/text/regex/RegexExtensions.kt",
    "src/kotlin/CharCode.kt",
    "src/kotlin/Boolean.kt",
    "src/kotlin/Nothing.kt",
    "src/kotlin/Enum.kt",
    "src/kotlin/Number.kt",
    "src/kotlin/Comparable.kt",
    "src/kotlin/Iterator.kt",
    "src/kotlin/CharSequence.kt",
    "src/kotlin/Any.kt",
    "src/kotlin/experimental/ExpectRefinement.kt",
    "src/kotlin/experimental/ExperimentalObjCEnum.kt",
    "src/kotlin/experimental/ExperimentalObjCName.kt",
    "src/kotlin/experimental/ExperimentalObjCRefinement.kt",
    "src/kotlin/reflect/KVariance.kt",
    "src/kotlin/reflect/KTypeProjection.kt",
    "src/kotlin/reflect/KClassifier.kt",
    "src/kotlin/reflect/KTypeParameter.kt",
    "src/kotlin/reflect/KType.kt",
    "src/kotlin/reflect/KCallable.kt",
    "src/kotlin/reflect/KFunction.kt",
    "src/kotlin/reflect/KProperty.kt",
    "src/kotlin/reflect/KClass.kt",
    "src/kotlin/reflect/KClasses.kt",
    "src/kotlin/reflect/typeOf.kt",
    "src/kotlin/Array.kt",
    "src/kotlin/Collections.kt",
    "src/kotlin/String.kt",
    "src/kotlin/Char.kt",
    "src/kotlin/Arrays.kt",
    "src/kotlin/ArrayIntrinsics.kt",
    "common/src/generated/_OneToManyTitlecaseMappings.kt",
    "common/src/generated/_Comparisons.kt",
    "common/src/generated/_UCollections.kt",
    "common/src/generated/_UComparisons.kt",
    "common/src/generated/_URanges.kt",
    "common/src/generated/_USequences.kt",
    "common/src/generated/_Arrays.kt",
    "common/src/generated/_UArrays.kt",
    "common/src/kotlin/ExceptionsH.kt",
    "common/src/kotlin/MathH.kt",
    "common/src/kotlin/TextH.kt",
    "common/src/kotlin/SequencesH.kt",
    "common/src/kotlin/collections/CollectionsH.kt",
    "src/kotlin/coroutines/cancellation/CancellationExceptionH.kt",
    "src/kotlin/collections/ArrayDeque.kt",
    "src/kotlin/collections/UArraySorting.kt",
    "src/kotlin/collections/Sequences.kt",
    "src/kotlin/ranges/PrimitiveRanges.kt",
    "src/kotlin/util/KotlinVersion.kt",
    "src/kotlin/collections/PrimitiveIterators.kt",
    "src/kotlin/collections/Arrays.kt",
    "common/src/kotlin/JsAnnotationsH.kt",
    "common/src/kotlin/JvmAnnotationsH.kt",
    "src/kotlin/annotations/NativeAnnotations.kt",
    "src/kotlin/annotations/NativeConcurrentAnnotations.kt",
    "src/kotlin/Primitives.kt",
    "src/kotlin/random/Random.kt",
    "src/kotlin/random/XorWowRandom.kt",
    "src/kotlin/text/HexFormat.kt",
    "src/kotlin/text/CharCategory.kt",
    "src/kotlin/collections/AbstractIterator.kt",
    "src/kotlin/text/Appendable.kt",
    "src/kotlin/text/Char.kt",
    "src/kotlin/text/StringNumberConversions.kt",
    "src/kotlin/text/Strings.kt",
    "src/kotlin/text/Indent.kt",
    "src/kotlin/text/Typography.kt",
    "src/kotlin/contracts/ContractBuilder.kt",
    "src/kotlin/contracts/Effect.kt",
    "src/kotlin/annotation/Annotations.kt",
    "src/kotlin/Annotation.kt",
    "src/kotlin/Annotations.kt",
    "src/kotlin/internal/Annotations.kt",
    "src/kotlin/internal/AnnotationsBuiltin.kt",
    "src/kotlin/internal/progressionUtil.kt",
    "src/kotlin/internal/serializationUtil.kt",
    "src/kotlin/experimental/bitwiseOperations.kt",
    "src/kotlin/experimental/inferenceMarker.kt",
];

/// klio-authored platform `actual` source files shipped in the same
/// `SOURCES` section, paths relative to the `klio-stdlib` crate root.
const KLIO_STDLIB_ACTUAL_FILES: &[&str] = &[
    "kotlin-time/Actuals.kt",
    "kotlin-coroutines/Actuals.kt",
    "kotlin-coroutines/Intrinsics.kt",
    "kotlin-collections/CollectionsActuals.kt",
    "kotlin-io/Closeable.kt",
    "kotlin-io/Serializable.kt",
    "kotlin-util/Lazy.kt",
];

/// Locate the local upstream Kotlin checkout's `libraries/stdlib`
/// directory (the same one the stdlib-gen miner and the parity harness
/// depend on). The workspace root is two levels up from the
/// `klio-stdlib` crate dir.
fn upstream_stdlib_root() -> Result<PathBuf, PackError> {
    let crate_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let ws_root = crate_dir
        .parent()
        .and_then(Path::parent)
        .ok_or_else(|| PackError::Io("cannot locate workspace root from klio-stdlib".into()))?;
    let root = ws_root.join("kotlin").join("libraries").join("stdlib");
    if !root.is_dir() {
        return Err(PackError::Io(format!(
            "upstream Kotlin checkout missing: expected stdlib sources at {} \
             (the curated stdlib SOURCES path and the parity harness both \
             require the local `kotlin/` checkout)",
            root.display()
        )));
    }
    Ok(root)
}

/// Read the curated upstream commonMain files plus the klio-authored
/// `actual` files into a [`SourceBundle`]. Fails hard (matching the
/// parity harness's hard dependency on the checkout) when an expected
/// file is absent.
fn build_curated_sources() -> Result<SourceBundle, PackError> {
    let upstream = upstream_stdlib_root()?;
    let crate_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let mut files: Vec<SourceFile> = Vec::new();

    for rel in CURATED_UPSTREAM_SOURCES {
        let abs = upstream.join(rel);
        let bytes = std::fs::read(&abs).map_err(|e| {
            PackError::Io(format!(
                "curated stdlib source {} unreadable: {e}",
                abs.display()
            ))
        })?;
        files.push(SourceFile {
            rel_path: format!("stdlib/kotlin/libraries/stdlib/{rel}"),
            bytes,
        });
    }

    for rel in KLIO_STDLIB_ACTUAL_FILES {
        let abs = crate_dir.join(rel);
        let bytes = std::fs::read(&abs).map_err(|e| {
            PackError::Io(format!(
                "klio stdlib actual source {} unreadable: {e}",
                abs.display()
            ))
        })?;
        files.push(SourceFile {
            rel_path: format!("stdlib/klio/{rel}"),
            bytes,
        });
    }

    Ok(SourceBundle { files })
}

/// Build a deterministic pack for the in-process Kotlin standard
/// library. When `compress_symbols` is true the symbol index section
/// is zstd-compressed (default for the embedded build).
pub fn build_stdlib_pack(compress_symbols: bool) -> Result<Vec<u8>, PackError> {
    let manifest = PackManifest {
        library_id: "stdlib".into(),
        library_version: env!("CARGO_PKG_VERSION").into(),
        abi_version: 1,
        implicit_packages: IMPLICITLY_IMPORTED_PACKAGES
            .iter()
            .map(|s| (*s).to_string())
            .collect(),
        dependencies: vec![],
    };
    let manifest_bytes = encode(&manifest)?;

    let mut sym_entries: Vec<SymbolRecord> = generated::stdlib_symbols()
        .iter()
        .map(symbol_entry_to_record)
        .collect();
    sym_entries.sort_by(|a, b| a.fqn.cmp(&b.fqn));
    let symbol_bytes = encode(&SymbolIndex { entries: sym_entries })?;

    let mut bindings: Vec<Binding> = Vec::new();
    let mut seen = std::collections::BTreeSet::<String>::new();
    for fqn in crate::all_symbol_names() {
        if implementation(fqn).is_none() {
            continue;
        }
        if !seen.insert(fqn.to_string()) {
            continue;
        }
        let arity = param_names(fqn).map(<[&str]>::len).unwrap_or(0);
        let max_arity: u8 = u8::try_from(arity).unwrap_or(u8::MAX);
        bindings.push(Binding {
            fqn: fqn.to_string(),
            kind: BindingKind::Function,
            host_symbol: fqn.to_string(),
            overrides_interpreter: true,
            purity: Purity::Effectful,
            min_arity: max_arity,
            max_arity,
            platform_actual: false,
        });
    }
    bindings.sort_by(|a, b| a.fqn.cmp(&b.fqn));
    let binding_bytes = encode(&BindingManifest { bindings })?;

    // Curated upstream commonMain + klio actuals, shipped so the loader
    // can parse + register them when a program imports a package these
    // sources provide (today: `kotlin.time`). The mined SYMBOLS /
    // BINDINGS sections above are still what gets statically linked;
    // this SOURCES section carries the Kotlin the interpreter parses.
    let sources_bytes = encode(&build_curated_sources()?)?;

    let mut writer = PackWriter::new();
    writer.add_raw(section_names::MANIFEST, manifest_bytes);
    writer.add_section(
        section_names::SYMBOLS,
        symbol_bytes,
        if compress_symbols { Compression::Zstd } else { Compression::None },
    );
    writer.add_raw(section_names::BINDINGS, binding_bytes);
    writer.add_section(section_names::SOURCES, sources_bytes, Compression::Zstd);
    writer.finish()
}

fn symbol_entry_to_record(e: &crate::SymbolEntry) -> SymbolRecord {
    let kind = match e.kind {
        crate::SymbolKind::Function => SymbolKind::Function,
        crate::SymbolKind::Property => SymbolKind::Property,
        crate::SymbolKind::Class => SymbolKind::Class,
        crate::SymbolKind::Interface => SymbolKind::Interface,
        crate::SymbolKind::Object => SymbolKind::Object,
        crate::SymbolKind::TypeAlias => SymbolKind::TypeAlias,
    };
    let modifiers = ModifierBits::from_bits_truncate(e.modifiers.0);
    SymbolRecord {
        fqn: e.fqn.to_string(),
        package: e.package.to_string(),
        name: e.name.to_string(),
        kind,
        receiver: e.receiver.map(str::to_string),
        signature: e.signature.to_string(),
        param_names: e.param_names.iter().map(|s| (*s).to_string()).collect(),
        modifiers,
        source: Some(SourceLoc {
            path: e.source.path.to_string(),
            line: e.source.line,
            column: e.source.column,
        }),
    }
}

File: src/stdlib/pack_builder.zig
//! Build a `.klio-pack` byte stream describing the stdlib surface. Used by
//! `klio pack stdlib` and by the embedded-pack build path.
//!
//! Emits `manifest`, `symbols`, `bindings`, and `sources` sections. The
//! Zig std has no zstd encoder, so every section is stored uncompressed; the
//! `compress_symbols` flag is accepted for source compatibility but has no
//! effect.

const std = @import("std");
const pack = @import("pack");

const root = @import("stdlib.zig");
const schema = pack.schema;
const PackError = pack.PackError;
const PackWriter = pack.PackWriter;
const section_names = pack.section_names;

/// The library version embedded in the pack manifest. The Rust build stamped
/// this from `CARGO_PKG_VERSION`; here it is the in-tree stdlib version.
pub const LIBRARY_VERSION: []const u8 = "0.1.0";

/// The stdlib source manifest (curated upstream files, klio actuals, and
/// their repo-relative roots) lives in `stdlib_sources.zig` so the top-level
/// build.zig can import the same lists; re-exported here for consumers.
pub const stdlib_sources = @import("stdlib_sources.zig");
pub const CURATED_UPSTREAM_SOURCES = stdlib_sources.CURATED_UPSTREAM_SOURCES;
pub const KLIO_STDLIB_ACTUAL_FILES = stdlib_sources.KLIO_STDLIB_ACTUAL_FILES;
pub const UPSTREAM_STDLIB_ROOT = stdlib_sources.UPSTREAM_STDLIB_ROOT;
pub const KLIO_STDLIB_DIR = stdlib_sources.KLIO_STDLIB_DIR;

/// Build a deterministic pack for the in-process Kotlin standard library.
///
/// `compress_symbols` is accepted for source compatibility with the Rust
/// builder; this build has no zstd encoder so every section is stored
/// uncompressed. On failure `result` is set and `null` is returned.
pub fn buildStdlibPack(
    allocator: std.mem.Allocator,
    compress_symbols: bool,
    result: *PackError,
) std.mem.Allocator.Error!?std.ArrayList(u8) {
    _ = compress_symbols;
    var arena_state = std.heap.ArenaAllocator.init(allocator);
    defer arena_state.deinit();
    const a = arena_state.allocator();

    // -- manifest --
    var implicit = try a.alloc([]const u8, root.IMPLICITLY_IMPORTED_PACKAGES.len);
    for (root.IMPLICITLY_IMPORTED_PACKAGES, 0..) |p, i| implicit[i] = p;
    const manifest = schema.PackManifest{
        .library_id = "stdlib",
        .library_version = LIBRARY_VERSION,
        .abi_version = 1,
        .implicit_packages = implicit,
        .dependencies = &.{},
        .default_features = &.{},
        .features = &.{},
    };
    const manifest_bytes = (try schema.encode(schema.PackManifest, a, &manifest, result)) orelse return null;

    // -- symbols --
    const syms = root.generated.stdlibSymbols();
    var sym_entries = try a.alloc(schema.SymbolRecord, syms.len);
    for (syms, 0..) |*e, i| sym_entries[i] = symbolEntryToRecord(e);
    std.mem.sort(schema.SymbolRecord, sym_entries, {}, lessRecordByFqn);
    const symbol_index = schema.SymbolIndex{ .entries = sym_entries };
    const symbol_bytes = (try schema.encode(schema.SymbolIndex, a, &symbol_index, result)) orelse return null;

    // -- bindings --
    var bindings: std.ArrayList(schema.Binding) = .empty;
    var seen = std.StringHashMap(void).init(a);
    var names = root.allSymbolNames();
    while (names.next()) |fqn| {
        if (root.implementation(fqn) == null) continue;
        const gop = try seen.getOrPut(fqn);
        if (gop.found_existing) continue;
        const arity: usize = if (root.paramNames(fqn)) |p| p.len else 0;
        const max_arity: u8 = std.math.cast(u8, arity) orelse std.math.maxInt(u8);
        try bindings.append(a, .{
            .fqn = fqn,
            .kind = .Function,
            .host_symbol = fqn,
            .overrides_interpreter = true,
            .purity = .Effectful,
            .min_arity = max_arity,
            .max_arity = max_arity,
            .platform_actual = false,
        });
    }
    std.mem.sort(schema.Binding, bindings.items, {}, lessBindingByFqn);
    const binding_manifest = schema.BindingManifest{ .bindings = bindings.items };
    const binding_bytes = (try schema.encode(schema.BindingManifest, a, &binding_manifest, result)) orelse return null;

    // -- sources --
    const sources = (try buildCuratedSources(a, result)) orelse return null;
    const sources_bytes = (try schema.encode(schema.SourceBundle, a, &sources, result)) orelse return null;

    // -- assemble --
    var writer = PackWriter.init(allocator);
    defer writer.deinit();
    _ = try writer.addRaw(section_names.MANIFEST, manifest_bytes.items);
    _ = try writer.addRaw(section_names.SYMBOLS, symbol_bytes.items);
    _ = try writer.addRaw(section_names.BINDINGS, binding_bytes.items);
    _ = try writer.addRaw(section_names.SOURCES, sources_bytes.items);
    return try writer.finish(result);
}

/// Read the curated upstream commonMain files plus the klio-authored `actual`
/// files into a `SourceBundle`. Fails as data when an expected file is absent.
fn buildCuratedSources(a: std.mem.Allocator, result: *PackError) std.mem.Allocator.Error!?schema.SourceBundle {
    var threaded: std.Io.Threaded = .init(a, .{});
    defer threaded.deinit();
    const io = threaded.io();
    const cwd = std.Io.Dir.cwd();
    var files: std.ArrayList(schema.SourceFile) = .empty;

    var upstream = cwd.openDir(io, UPSTREAM_STDLIB_ROOT, .{}) catch {
        result.* = .{ .Io = "upstream Kotlin checkout missing: expected stdlib sources at kotlin/libraries/stdlib" };
        return null;
    };
    defer upstream.close(io);

    for (CURATED_UPSTREAM_SOURCES) |rel| {
        var bytes: []const u8 = upstream.readFileAlloc(io, rel, a, .unlimited) catch |e| switch (e) {
            error.OutOfMemory => return error.OutOfMemory,
            else => {
                result.* = .{ .Io = "curated stdlib source unreadable" };
                return null;
            },
        };
        if (std.mem.eql(u8, rel, stdlib_sources.KOTLIN_VERSION_FILE)) {
            bytes = (try stampKotlinVersion(a, bytes)) orelse {
                result.* = .{ .Io = "KotlinVersion placeholder not found; review KOTLIN_RELEASE in stdlib_sources.zig against the pinned kotlin/ tag" };
                return null;
            };
        }
        const rel_path = try std.fmt.allocPrint(a, "stdlib/kotlin/libraries/stdlib/{s}", .{rel});
        try files.append(a, .{ .rel_path = rel_path, .bytes = bytes });
    }

    var klio_dir = cwd.openDir(io, KLIO_STDLIB_DIR, .{}) catch {
        result.* = .{ .Io = "klio-stdlib directory missing: expected actuals at kotlin-klio" };
        return null;
    };
    defer klio_dir.close(io);

    for (KLIO_STDLIB_ACTUAL_FILES) |rel| {
        const bytes = klio_dir.readFileAlloc(io, rel, a, .unlimited) catch |e| switch (e) {
            error.OutOfMemory => return error.OutOfMemory,
            else => {
                result.* = .{ .Io = "klio stdlib actual source unreadable" };
                return null;
            },
        };
        const rel_path = try std.fmt.allocPrint(a, "stdlib/klio/{s}", .{rel});
        try files.append(a, .{ .rel_path = rel_path, .bytes = bytes });
    }

    return .{ .files = files.items };
}

/// Rewrites the `KotlinVersion(major, minor, 255)` build placeholder in
/// `KotlinVersion.kt` to the pinned release triple, mirroring the rewrite
/// kotlinc's build performs on the same source. Returns null when the
/// placeholder is absent (a version bump changed the file shape).
fn stampKotlinVersion(a: std.mem.Allocator, bytes: []const u8) std.mem.Allocator.Error!?[]const u8 {
    const needle = stdlib_sources.KOTLIN_VERSION_PLACEHOLDER;
    const idx = std.mem.indexOf(u8, bytes, needle) orelse return null;
    var out: std.ArrayList(u8) = .empty;
    try out.appendSlice(a, bytes[0..idx]);
    try out.appendSlice(a, stdlib_sources.KOTLIN_VERSION_STAMPED);
    try out.appendSlice(a, bytes[idx + needle.len ..]);
    return try out.toOwnedSlice(a);
}

test stampKotlinVersion {
    const a = std.testing.allocator;
    const src = "fun get(): KotlinVersion = KotlinVersion(2, 4, 255) // stamped";
    const got = (try stampKotlinVersion(a, src)).?;
    defer a.free(got);
    try std.testing.expectEqualStrings(
        "fun get(): KotlinVersion = KotlinVersion(2, 4, 0) // stamped",
        got,
    );
    try std.testing.expectEqual(@as(?[]const u8, null), try stampKotlinVersion(a, "no placeholder here"));
}

fn symbolEntryToRecord(e: *const root.SymbolEntry) schema.SymbolRecord {
    const kind: schema.SymbolKind = switch (e.kind) {
        .Function => .Function,
        .Property => .Property,
        .Class => .Class,
        .Interface => .Interface,
        .Object => .Object,
        .TypeAlias => .TypeAlias,
    };
    return .{
        .fqn = e.fqn,
        .package = e.package,
        .name = e.name,
        .kind = kind,
        .receiver = e.receiver,
        .signature = e.signature,
        .param_names = @constCast(e.param_names),
        .modifiers = schema.ModifierBits.fromBits(e.modifiers.bits),
        .source = .{ .path = e.source.path, .line = e.source.line, .column = e.source.column },
    };
}

fn lessRecordByFqn(_: void, x: schema.SymbolRecord, y: schema.SymbolRecord) bool {
    return std.mem.order(u8, x.fqn, y.fqn) == .lt;
}

fn lessBindingByFqn(_: void, x: schema.Binding, y: schema.Binding) bool {
    return std.mem.order(u8, x.fqn, y.fqn) == .lt;
}

const testing = std.testing;

test "symbol entry to record preserves fqn and kind" {
    const params = [_][]const u8{"message"};
    const e = root.SymbolEntry{
        .fqn = "kotlin.io.println",
        .package = "kotlin.io",
        .name = "println",
        .kind = .Function,
        .receiver = null,
        .signature = "public fun println(message: Any?): Unit",
        .param_names = &params,
        .modifiers = .{ .bits = root.Modifiers.PUBLIC },
        .source = .{ .path = "Console.kt", .line = 42, .column = 1 },
        .impl_fn = null,
    };
    const r = symbolEntryToRecord(&e);
    try testing.expectEqualStrings("kotlin.io.println", r.fqn);
    try testing.expectEqual(schema.SymbolKind.Function, r.kind);
    try testing.expect(r.modifiers.PUBLIC);
    try testing.expectEqual(@as(usize, 1), r.param_names.len);
}
